fix(nuget): add nuget dependency extract

hellhound-abu opened the pull request:
<!-- Delete Template sections if unneccesary -->
<!-- Add issue number here (We encourage you to create the Issue First) -->
<!-- You can also link the issue in Commit Messages -->

Fixes #0000

<!--
Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!
* [x] - Checked Box
* [ ] - Unchecked Box
-->

### Tasks

* [ ] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [ ] PR is descriptively titled 📑 and links the original issue above 🔗
* [ ] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [ ] Commits are in uniquely-named feature branch and has no merge conflicts 📁

<!--
We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged**
if your tests fail at first!
If tests do fail, click on the red `X` to learn why by reading the logs.
Thanks!
-->

<!-- Don't forget to Signoff -->

bloodhoundGit committed:
fix(nuget): add nuget dependency extract

Signed-off-by: Abedin Poonawala <abedinp@securestack.com>

AyanSinhaMahapatra commented:
Thanks! Could you fix the test failures at `tests/packagedcode/test_nuget.py` with:

* `pytest -vvs tests/packagedcode/test_nuget.py` and then
* `SCANCODE_REGEN_TEST_FIXTURES=yes pytest -vvs tests/packagedcode/test_nuget.py --lf`

Also please add your signoff. :bow:

pombredanne commented:
@hellhound-abu gentle ping.. we would like to merge this... but we cannot unless you update your PR at least with a DCO signoff.

pombredanne commented:
@6mile ping you seem to be from the same group, may be you can help there?

6mile commented:
> @6mile ping you seem to be from the same group, may be you can help there?

I've let him know.  Cheers.

jaitaiwan commented:
Hey folks, Dan from SecureStack here. Abu is on holidays so won’t be able to contribute till 2 weeks from now. I can fix the failing tests if that’s helpful but that won’t help get a DCO from him.

Do you have a way around this or content to wait?

pombredanne commented:
@jaitaiwan Thank you ++ for your reply! 
No rush,  we can wait alright!
We are just making sure we do not have too many PRs that could go stale from lack of TLC especially like this one which is valuable addition.
 
Whatever you guys are doing on SecureStack looks great BTW! :+1:

pombredanne commented:
@hellhound-abu gentle ping ... you may be back now?

bloodhoundGit committed:
feat(nuget): add dependency check for inside group tags

    1. update code to look for dependencies in a group as well
    2. update nuget test data with dependency data

Signed-off-by: Abedin Poonawala <abedinp@securestack.com>

bloodhoundGit committed:
fix(nuget): match updated develop code

Signed-off-by: Abedin Poonawala <abedinp@securestack.com>

bloodhoundGit committed:
fix(nuget): update nuget test

Signed-off-by: Abedin Poonawala <abedinp@securestack.com>

bloodhoundGit committed:
fix(nuget): add dependency section in bootstrap test

Signed-off-by: Abedin Poonawala <abedinp@securestack.com>

hellhound-abu commented:
Updated the branch with necessary changes. If we can merge it whenever possible.

Thanks for waiting.

pombredanne commented on src/packagedcode/nuget.py:
```diff
@@ -12,10 +12,92 @@
 """
-# TODO: add dependencies
+
+# TODO: Add section to handle dependencies of dependencies
```
For this you may want to create a separate issue... TODO comments (that I tend to abuse) eventually get lost.
Note also that for a full, dynamic dependency resolution of nugets. there is this new project at https://github.com/nexB/nuget-inspector

jaitaiwan replied:
I say the same thing about TODO's internally 😂 That nuget-inspector proejct looks promising.

pombredanne commented on src/packagedcode/nuget.py:
```diff
@@ -12,10 +12,92 @@
+                            type='nuget',
+                            namespace=None,
+                            name=dependency.get("@id"),
+                            version=dependency.get("@version"),
```
You can only put a version here if this is not a range so IMHO you would have to test for this.

pombredanne commented on src/packagedcode/nuget.py:
```diff
@@ -12,10 +12,92 @@
+                            dep_pack = models.DependentPackage(
+                            purl=str(dpurl),
+                            extracted_requirement=dependency.get("@version"),
+                            scope="dependency",
```
Would not each dependency group become a scope here?

pombredanne commented on src/packagedcode/nuget.py:
```diff
@@ -12,10 +12,92 @@
+                            scope="dependency",
+                            is_runtime=False,
+                            is_optional=False,
+                            is_resolved=True,
```
A dependency is resolved if an only if we have a concrete version and not a range. Here the version is optional AFAIK and could be a range?

pombredanne commented on src/packagedcode/nuget.py:
```diff
@@ -12,10 +12,92 @@
+
+                            dependencies.append(dep_pack)
+                            
+            if "dependency" in nuspec.get("dependencies"):
```
I would prefer something like this, and then reuse the `nuget_dependency` var below:
```suggestion
            nuget_dependencies = nuspec.get("dependencies") or {}
            nuget_dependency = nuget_dependencies.get("dependency")
            if nuget_dependency:
```

jaitaiwan replied:
Out of curiosity, any reason for `or {}` over `get("...", {})`?

pombredanne replied:
@jaitaiwan re:

> Out of curiosity, any reason for or {} over get("...", {})?

This is an idiom I like to use to make sure that I am getting a proper empty dict in all cases even if the get returns None or some emptyish value.
Getting an empty dict when you expect a dict enable chaining the next call without requiring an extra check for empties.

```Python
>>> repr({1: None, 2: "", 3: {"bar": "baz"}}.get(1, {}))
'None'
>>> repr({1: None, 2: "", 3: {"bar": "baz"}}.get(1, {}) or {})
'{}'
>>> repr({1: None, 2: "", 3: {"bar": "baz"}}.get(1) or {})
'{}'

```

jaitaiwan replied:
Fair call. thanks!

pombredanne commented on src/packagedcode/nuget.py:
```diff
@@ -12,10 +12,92 @@
+                    dep_pack = models.DependentPackage(
+                        purl=str(dpurl),
+                        extracted_requirement=nuspec.get("dependencies").get(
+                            "dependency").get("@version"),
```
Factor version here and above in a variable.

pombredanne commented on src/packagedcode/nuget.py:
```diff
@@ -12,10 +12,92 @@
+                        scope="dependency",
+                        is_runtime=False,
+                        is_optional=False,
+                        is_resolved=True,
```
Same comment as above... resolve is True IFF we have a concrete pinned, non-empty version

pombredanne commented on src/packagedcode/nuget.py:
```diff
@@ -12,10 +12,92 @@
+        return dependencies
+
+    except Exception as e:
+        print(e)
```
Please do not print exceptions. In general your should avoid to wrap you code in a try/except altogether as this can never be debugged this way. Let the exception bubble up and we will catch it where needed, and worse case at the top level.

pombredanne commented on src/packagedcode/nuget.py:
```diff
@@ -62,14 +144,15 @@ def parse(cls, location):
         # Summary: A short description of the package for UI display. If omitted, a
         # truncated version of description is used.
-        description = build_description(nuspec.get('summary') , nuspec.get('description'))
+        description = build_description(
```
Why wrapping this on tow lines? I am not sure this helps with readability.

pombredanne commented on tests/packagedcode/data/nuget/bootstrap.nuspec.json.expected:
```diff
@@ -1,104 +1,115 @@
-    "datasource_id": "nuget_nupsec",
-    "purl": "pkg:nuget/bootstrap@4.0.0-alpha"
-  }
+          "type": "nuget",
```
Please avoid changing the spacing if possible as this makes reviewing the diff difficult

pombredanne reviewed (changes requested):
@hellhound-abu Thank you ++ for the updates! I posted a few review comments inline for your consideration. There is also one failing test.

pombredanne commented:
@hellhound-abu Happy new year! gentle ping

pombredanne commented:
@AyanSinhaMahapatra since there is no reply for now, let's move this to the next milestone

AyanSinhaMahapatra commented:
@pombredanne moved this to v32.1

pombredanne commented:
I have applied the fixes in https://github.com/nexB/scancode-toolkit/pull/3206 on top of your commits. So I am losing this and will merge this other branch!
Thanks!
